sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala

```diff
@@ -46,15 +48,24 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   /**
    * An execution strategy for rules that indicates the maximum number of executions. If the
    * execution reaches fix point (i.e. converge) before maxIterations, it will stop.
+   * If throwsExceptionUponMaxIterations is equal to true, it will issue an exception
+   * TreeNodeException.
    */
-  abstract class Strategy { def maxIterations: Int }
+  abstract class Strategy {
+    def maxIterations: Int
+    def throwsExceptionUponMaxIterations: Boolean
+  }
 
   /** A strategy that only runs once. */
-  case object Once extends Strategy { val maxIterations = 1 }
+  case object Once extends Strategy {
+    override val maxIterations = 1
+    override val throwsExceptionUponMaxIterations = false
```

Comment:
Isn't this kind of janky to insert a field into the non-test code just to get the behavior you want in a test? how about mocking subclasses?